File: python/meetup/meetup.py
import datetime


class MeetupDayException(Exception):
    pass


def meetup(year, month, week, weekday):
    """Determine a meetup day from a year, month, day, and vague week.

    :param year: int - the year for the meetup.
    :param month: int - the month for the meetup.
    :param week: str - a vague delineater for the week of the meetup.
    :param weekday: str - the day of the week for the meetup.

    :return: date - the meetup date.

    :raises MeetupDayException: That day does not exist.
    """
    intervals = {
        "teenth": 13,
        "first": 1,
        "second": 8,
        "third": 15,
        "fourth": 22,
        "fifth": 29,
        "last": 22,
    }

    try:
        approximate_day = datetime.date(year, month, intervals[week])
        refined_day = _match_day(approximate_day, weekday)
        return _match_week(refined_day, week)
    except ValueError:
        raise MeetupDayException("That day does not exist.")


def _match_day(approximate_day: datetime.date, day: str) -> datetime.date:
    weekdays = (
        "Monday",
        "Tuesday",
        "Wednesday",
        "Thursday",
        "Friday",
        "Saturday",
        "Sunday",
    )
    while weekdays[approximate_day.weekday()] != day:
        approximate_day = datetime.date(
            approximate_day.year, approximate_day.month, approximate_day.day + 1
        )
    return approximate_day


def _match_week(date_obj: datetime.date, interval: int) -> datetime.date:
    if (
        date_obj.month != 2
        and (interval == "last" and date_obj.day < 25)
        or date_obj.month == 2
        and interval == "last"
        and is_leap_year(date_obj.year)
    ):
        date_obj += datetime.timedelta(days=7)
    return date_obj


def is_leap_year(year: int) -> bool:
    return (year % 100 != 0 and year % 4 == 0) or year % 400 == 0

File: ruby/meetup/meetup.rb
require 'date'

# Determine's a meetup day from a year, month, day, and vague week.
class Meetup
  class InvalidDay < RuntimeError; end

  MONTH_INTERVALS = {
    teenth: 13,
    first: 1,
    second: 8,
    third: 15,
    fourth: 22,
    fifth: 29,
    last: 25,
    feb_last: 23
  }.freeze

  def initialize(month, year)
    @month = month
    @year = year
  end

  def day(weekday, week)
    week = :feb_last if last_week_of_february?(week)
    initial = Date.new(@year, @month, MONTH_INTERVALS[week])
    match_day(initial, weekday)
  rescue
    raise InvalidDay('Oh, what a day.')
  end

  private

  def last_week_of_february?(week)
    @month == 2 && week == :last
  end

  def match_day(initial_date, weekday)
    return initial_date if initial_date.send(weekday.to_s + '?')
    match_day(initial_date.next, weekday)
  end
end
